File: rental/src/main/java/io/axoniq/demo/bikerental/rental/support/AxonServerClient.java
package io.axoniq.demo.bikerental.rental.support;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientResponseException;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Thin client over the Axon Server <em>Integration</em> HTTP API. This is the HTTP equivalent of the
 * gRPC branch's command/query/event channels: it is used to send commands and queries, publish and
 * read events, and to register this application's handler endpoints so Axon Server pushes messages
 * to them.
 * <p>
 * All calls target a single {@code context} (default) and are plain blocking HTTP, so callers run
 * them off the request/event-loop threads (see {@code CommandDispatcher}, the handler controllers).
 */
@Component
public class AxonServerClient {

    private static final Logger logger = LoggerFactory.getLogger(AxonServerClient.class);
    // The media type the Axon Server event-publish endpoint consumes (per its OpenAPI definition).
    private static final MediaType EVENT_LIST_JSON = MediaType.parseMediaType("application/vnd.axoniq.event.list+json");

    private final RestClient restClient;
    private final String context;

    public AxonServerClient(@Value("${axon.axonserver.http-url:http://localhost:8024}") String axonServerUrl,
                            @Value("${axon.axonserver.context:default}") String context) {
        this.restClient = RestClient.builder().baseUrl(axonServerUrl).build();
        this.context = context;
    }

    // ---------------------------------------------------------------------------------------------
    // Messaging: send commands / queries, publish & read events
    // ---------------------------------------------------------------------------------------------

    public AxonServerMessages.CommandResult sendCommand(AxonServerMessages.Command command) {
        return restClient.post()
                         .uri(uri -> uri.path("/v2/commands").queryParam("context", context).build())
                         .contentType(MediaType.APPLICATION_JSON)
                         .body(command)
                         .retrieve()
                         .onStatus(status -> status.isError(), (req, res) -> {
                             throw new CommandExecutionException(
                                     "AXONIQ-HTTP", readError(res.getStatusText(), res.getBody()));
                         })
                         .body(AxonServerMessages.CommandResult.class);
    }

    public AxonServerMessages.QueryResult sendQuery(AxonServerMessages.Query query) {
        return restClient.post()
                         .uri(uri -> uri.path("/v2/queries").queryParam("context", context).build())
                         .contentType(MediaType.APPLICATION_JSON)
                         .body(query)
                         .retrieve()
                         .onStatus(status -> status.isError(), (req, res) -> {
                             throw new QueryExecutionException(readError(res.getStatusText(), res.getBody()));
                         })
                         .body(AxonServerMessages.QueryResult.class);
    }

    /** Appends events in a single transaction (POST /v2/events with the wrapped JSON list form). */
    public void appendEvents(List<AxonServerMessages.Event> events) {
        if (events.isEmpty()) {
            return;
        }
        restClient.post()
                  .uri(uri -> uri.path("/v2/events").queryParam("context", context).build())
                  .contentType(EVENT_LIST_JSON)
                  .body(events)
                  .retrieve()
                  .toBodilessEntity();
    }

    /**
     * Reads an aggregate's event stream (ascending sequence), used to rebuild a decision model.
     * A not-yet-existing aggregate (no events) yields an empty list.
     */
    public List<AxonServerMessages.Event> readAggregateEvents(String aggregateId) {
        try {
            AxonServerMessages.Event[] events = restClient.get()
                    .uri(uri -> uri.path("/v2/aggregates/{id}/events")
                                   .queryParam("context", context)
                                   .build(aggregateId))
                    .retrieve()
                    .body(AxonServerMessages.Event[].class);
            return events == null ? List.of() : List.of(events);
        } catch (RestClientResponseException e) {
            if (e.getStatusCode().value() == 404) {
                return List.of();
            }
            throw e;
        }
    }

    // ---------------------------------------------------------------------------------------------
    // Integration registration: tell Axon Server where to push commands/queries/events
    // ---------------------------------------------------------------------------------------------

    /** Registers (or looks up an existing) endpoint and returns its id. */
    public String registerEndpoint(Map<String, Object> endpoint) {
        String name = (String) endpoint.get("name");
        Optional<String> existing = findEndpointId(name);
        if (existing.isPresent()) {
            logger.info("Integration endpoint '{}' already registered ({})", name, existing.get());
            return existing.get();
        }
        restClient.post()
                  .uri(uri -> uri.path("/v2/endpoints").queryParam("context", context).build())
                  .contentType(MediaType.APPLICATION_JSON)
                  .body(endpoint)
                  .retrieve()
                  .toBodilessEntity();
        return findEndpointId(name).orElseThrow(
                () -> new IllegalStateException("Endpoint '" + name + "' was not found after registration"));
    }

    public Optional<String> findEndpointId(String name) {
        List<Map<String, Object>> endpoints = restClient.get()
                .uri("/v2/endpoints")
                .retrieve()
                .body(new org.springframework.core.ParameterizedTypeReference<>() {
                });
        if (endpoints == null) {
            return Optional.empty();
        }
        return endpoints.stream()
                        .filter(e -> name.equals(e.get("name")) && context.equals(e.get("context")))
                        .map(e -> String.valueOf(e.get("id")))
                        .findFirst();
    }

    public void registerCommandHandler(String endpointId, Map<String, Object> handler) {
        registerHandler(endpointId, "commandHandlers", handler);
    }

    public void registerQueryHandler(String endpointId, Map<String, Object> handler) {
        registerHandler(endpointId, "queryHandlers", handler);
    }

    public void registerEventHandler(String endpointId, Map<String, Object> handler) {
        registerHandler(endpointId, "eventHandlers", handler);
    }

    private void registerHandler(String endpointId, String kind, Map<String, Object> handler) {
        try {
            restClient.post()
                      .uri(uri -> uri.path("/v2/endpoints/{endpoint}/" + kind)
                                     .queryParam("context", context)
                                     .build(endpointId))
                      .contentType(MediaType.APPLICATION_JSON)
                      .body(handler)
                      .retrieve()
                      .toBodilessEntity();
            logger.info("Registered {} '{}'", kind, handler.get("name"));
        } catch (RestClientResponseException e) {
            // Most likely the handler is already registered from a previous run - safe to ignore.
            logger.info("Skipping {} '{}' registration: {}", kind, handler.get("name"), e.getStatusText());
        }
    }

    private static String readError(String statusText, java.io.InputStream body) {
        try {
            byte[] bytes = body.readAllBytes();
            return bytes.length == 0 ? statusText : new String(bytes);
        } catch (Exception e) {
            return statusText;
        }
    }
}

File: core-api/src/support/axon-server-client.ts
import { createLogger } from './logger.js';
import { CommandExecutionException, QueryExecutionException } from './errors.js';
import type {
  Command,
  CommandResult,
  Event,
  Query,
  QueryResult,
} from './messages.js';

const logger = createLogger('AxonServerClient');

// The media type the Axon Server event-publish endpoint consumes (per its OpenAPI definition).
const EVENT_LIST_JSON = 'application/vnd.axoniq.event.list+json';

export interface AxonServerClientOptions {
  /** Axon Server's HTTP base URL, e.g. http://localhost:8024. */
  httpUrl: string;
  /** The context all calls target. */
  context: string;
}

/**
 * Thin client over the Axon Server Integration HTTP API: send commands and queries, publish and read
 * events, and register this application's handler endpoints so Axon Server pushes messages to them.
 * All calls target a single context and are plain HTTP.
 */
export class AxonServerClient {
  private readonly baseUrl: string;
  private readonly context: string;

  constructor(options: AxonServerClientOptions) {
    this.baseUrl = options.httpUrl.replace(/\/+$/, '');
    this.context = options.context;
  }

  // -------------------------------------------------------------------------------------------
  // Messaging: send commands / queries, publish & read events
  // -------------------------------------------------------------------------------------------

  async sendCommand(command: Command): Promise<CommandResult> {
    const response = await fetch(this.url('/v2/commands'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(command),
    });
    if (!response.ok) {
      throw new CommandExecutionException('AXONIQ-HTTP', await readError(response));
    }
    return (await response.json()) as CommandResult;
  }

  async sendQuery(query: Query): Promise<QueryResult> {
    const response = await fetch(this.url('/v2/queries'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(query),
    });
    if (!response.ok) {
      throw new QueryExecutionException(await readError(response));
    }
    return (await response.json()) as QueryResult;
  }

  /** Appends events in a single transaction (POST /v2/events with the wrapped JSON list form). */
  async appendEvents(events: Event[]): Promise<void> {
    if (events.length === 0) {
      return;
    }
    const response = await fetch(this.url('/v2/events'), {
      method: 'POST',
      headers: { 'Content-Type': EVENT_LIST_JSON },
      body: JSON.stringify(events),
    });
    if (!response.ok) {
      throw new Error(`Appending events failed: ${await readError(response)}`);
    }
  }

  /**
   * Reads an aggregate's event stream (ascending sequence), used to rebuild a decision model. A
   * not-yet-existing aggregate (no events) yields an empty list.
   */
  async readAggregateEvents(aggregateId: string): Promise<Event[]> {
    const response = await fetch(
      this.url(`/v2/aggregates/${encodeURIComponent(aggregateId)}/events`),
    );
    if (response.status === 404) {
      return [];
    }
    if (!response.ok) {
      throw new Error(`Reading aggregate ${aggregateId} failed: ${await readError(response)}`);
    }
    const events = (await response.json()) as Event[] | null;
    return events ?? [];
  }

  // -------------------------------------------------------------------------------------------
  // Integration registration: tell Axon Server where to push commands/queries/events
  // -------------------------------------------------------------------------------------------

  /** Registers (or looks up an existing) endpoint and returns its id. */
  async registerEndpoint(endpoint: Record<string, unknown>): Promise<string> {
    const name = String(endpoint.name);
    const existing = await this.findEndpointId(name);
    if (existing) {
      logger.info(`Integration endpoint '${name}' already registered (${existing})`);
      return existing;
    }
    const response = await fetch(this.url('/v2/endpoints'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(endpoint),
    });
    if (!response.ok) {
      throw new Error(`Registering endpoint '${name}' failed: ${await readError(response)}`);
    }
    const id = await this.findEndpointId(name);
    if (!id) {
      throw new Error(`Endpoint '${name}' was not found after registration`);
    }
    return id;
  }

  async findEndpointId(name: string): Promise<string | undefined> {
    const response = await fetch(this.url('/v2/endpoints'));
    if (!response.ok) {
      throw new Error(`Listing endpoints failed: ${await readError(response)}`);
    }
    const endpoints = (await response.json()) as Record<string, unknown>[] | null;
    return endpoints
      ?.find((e) => e.name === name && e.context === this.context)
      ?.id?.toString();
  }

  registerCommandHandler(endpointId: string, handler: Record<string, unknown>): Promise<void> {
    return this.registerHandler(endpointId, 'commandHandlers', handler);
  }

  registerQueryHandler(endpointId: string, handler: Record<string, unknown>): Promise<void> {
    return this.registerHandler(endpointId, 'queryHandlers', handler);
  }

  registerEventHandler(endpointId: string, handler: Record<string, unknown>): Promise<void> {
    return this.registerHandler(endpointId, 'eventHandlers', handler);
  }

  private async registerHandler(
    endpointId: string,
    kind: string,
    handler: Record<string, unknown>,
  ): Promise<void> {
    try {
      const response = await fetch(this.url(`/v2/endpoints/${endpointId}/${kind}`), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(handler),
      });
      if (!response.ok) {
        // Most likely the handler is already registered from a previous run - safe to ignore.
        logger.info(`Skipping ${kind} '${handler.name}' registration: ${response.statusText}`);
        return;
      }
      logger.info(`Registered ${kind} '${handler.name}'`);
    } catch (error) {
      logger.info(`Skipping ${kind} '${handler.name}' registration: ${String(error)}`);
    }
  }

  private url(path: string): string {
    return `${this.baseUrl}${path}?context=${encodeURIComponent(this.context)}`;
  }
}

async function readError(response: Response): Promise<string> {
  try {
    const body = await response.text();
    return body.length === 0 ? response.statusText : body;
  } catch {
    return response.statusText;
  }
}
